### ISSUE TITLE
Multiple values in a cell

### ISSUE THREAD

**Issue by kcoyle** [dcmi/dctap#4]
There are various situations where one may want to have multiple values in a cell, such as:

valueNodeType = LITERAL or BNODE
valueConstraint = red blue green (a pick list)
valueType = xsd:string or rdf:langString

How can this be done and still be [valid CSV](https://tools.ietf.org/html/rfc4180)?

[Original discussion in DCAP repo.](https://github.com/dcmi/dcap/issues/65)
[Other discussion in DCAP repo](https://github.com/dcmi/dcap/issues/37)

**Comment by kcoyle** [comment]
Note: there are some situations that we may not want to accommodate. For example if "valueNodeType is LITERAL or IRI" and you need to give a specific literal value type, that would require ingest programs to determine if the incoming value is a literal and then apply the literal value type, but not to apply the type to an IRI. (That's logical) But then if you also want to specify that if it is an IRI that IRI must have a specific namespace stem (e.g. https://schema.org/) then the degree of munging for the input program or the ingesting program is pretty messy. We need to figure out how far we can take this without overly complicating the feature.

**Comment by tombaker** [comment]
@kcoyle 
Another usecase would be a picklist of language tabs.

[RFC 4180](https://tools.ietf.org/html/rfc4180) treats cells simply as values, whereas we want to be able to split those values into component parts, which AFAICT is not addressed. I like this [simple answer on Stackoverflow](https://stackoverflow.com/questions/3853614/multiple-values-for-one-field-in-comma-separated-values-csv-format#:~:text=Yes.,entire%20field%20in%20double%20quotes.):
> You can have multiple values in one field of a CSV file.
> 
> If the multiple values are separated by space you need not do anything special but if the values are separated by a comma, you need to enclose the entire field in double quotes.

I expect we could probably all agree that spaces are a good way to separate the items of simple picklists and that, assuming we want to accommodate in-line picklists at all, we could recommend this as a way to do it.

I'm not as certain we would want to actually recommend the use of quotes and commas within a cell, because this quickly takes us into the territory of CSV dialects. The standard Python library for parsing CSV supports three dialects - "excel", "excel-tab", and "unix" - but I'm sure there are others. Rather than prescribe a specific way to format and parse such cells, I would prefer if we could simply point to some options and leave it to users to sort out exactly how to do it. There are many resources about such things on the Web.

One approach to which we could point is @johnsamuelwrites [ShExStatements](https://github.com/johnsamuelwrites/ShExStatements/) which, if I correctly recall, supports using the comma for separating values within a cell in cases where the CSV as a whole uses a vertical bar (instead of a comma) as the separator. In my own Python script, I would ideally like to support tab- and verticalbar-separated CSVs, though I think of such possibilities as implementation details that need not be covered in the DCTAP spec with anything more than a mention.

**Comment by kcoyle** [comment]
Proposed wording:

There are various situations where one may want to have multiple values in a cell, such as:

valueNodeType = IRI or BNODE
valueConstraint = red or blue or green 
valueType = xsd:string or rdf:langString

Multiple values in a cell are given as a list of whitespace-delimited values (commonly a space, a tab, or a line break) or a list delimited by another character that is defined in accompanying documentation. Multiple values in a cell are processed in a logical "or" relation. Thus the cell contents 

|     |
| ----|
|A B C| 

is processed as "A or B or C". In metadata creation applications this is often referred to as a "picklist".

When using whitespace as a delimiter, values that themselves contain whitespace will need to be enclosed using single or double quotes.

| propertyID | valueDatatype | valueConstraint | 
| ---- | ---- | ---- | 
| dct:subject | xsd:string | History Science Art | 

| propertyID | valueDatatype | valueConstraint |
| ---- | ---- | ---- | 
| dct:subject | xsd:string | "European History" Science "Fine Arts" | 

If appropriately documented, alternative delimiters between values can be used. These may obviate the need for quotes around values.

| propertyID | valueDatatype | valueConstraint | 
| ---- | ---- | ---- | 
| dct:subject | xsd:string | \European History\Science\Fine Arts |

**Comment by tombaker** [comment]
However, "History Science Art" is a perfectly good string and it seems counterintuitive that it would be parsed as three values simply because it contains whitespace. Rather, I think we should be saying that, in order to be parsed as a picklist, a valueConstraintType of Picklist should be used:

| propertyID | valueDatatype | valueConstraint | valueConstraintType|
| ---- | ---- | ---- | ---- |
| dct:subject | xsd:string | History Science Art | Picklist |


With regard to the use of punctuation for multiple values, I would prefer that we simply say:
- [as above] that multiple values are processed as "or" (though in principle, there might be value constraint types that process multiple values differently)
- [as above] that whitespace can be used to separate multiple values - as long as the values themselves do not contain whitespace
- that if values do have whitespace, follow the conventions used by your community, spreadsheet program, or other application
- that spreadsheets and other applications may have built-in support for punctuation and character escaping on the backend

In other words, it is good to say that whitespace will be interpreted as a value separator, but I think we should stop there. We can mention the use of commas but rather than illustrate how this is done, we can point off to CSV specifications or software documentation where the use of punctuation and escaping for multiple values is discussed.

**Comment by tombaker** [comment]
@philbarker With regard to the use of punctuation to separate multiple values, we might add, as Phil puts it in #5 : "Such mechanisms may not be interoperable."

**Comment by tombaker** [comment]
Maybe this can help us clarify which fields can have multiple values, and how.

| element | multiple values? | type |
| ---- | ---- | ---- | 
| shapeID | never | always URI |
| shapeLabel | maybe? | annotation (string) |
| propertyID | maybe? | always URI |
| propertyLabel | maybe? | annotation (string) |
| mandatory | never | boolean |
| repeatable | never | boolean |
| valueNodeType | yes | controlled string value |
| valueDataType | maybe | always URI |
| valueConstraint | yes | anything goes |
| valueConstraintType | never | controlled string value |
| valueShape | never | always URI |
| note | maybe? | annotation (string) |

**Comment by philbarker** [comment]
Why not multiple valueShapes? e.g. author could be Organization or Person shape.

**Comment by tombaker** [comment]
I typed the following into three cells of an Apple Numbers spreadsheet:
```
"red" "white" "blue"
"red", "white", "blue"
red, white, blue
```
and here's the CSV (with seven columns):
```
,"""red"" ""white"" ""blue""",,,,,
,"""red"", ""white"", ""blue""",,,,,
,"red, white, blue",,,,,
```

**Comment by philbarker** [comment]
For comparison, Google Sheets (without the extra columns):
```
"""red"" ""white"" ""blue"""
"""red"", ""white"", ""blue"""
"red, white, blue"
```

And LibreOffice Calc (with default delimiter options)
```
"""red"" ""white"" ""blue"""
"""red"", ""white"", ""blue"""
"red, white, blue"
```
and a random online [csv editor](https://edit-csv.net/):
```
"""red"" ""white"" ""blue""",,,,
"""red"", ""white"", ""blue""",,,,
"red, white, blue",,,,
```

As a final test I uploaded one of the CSV files into OntoRefine (just as an example of something other than a spreadsheet that allows CSV import) and got the same data back.

With python I have read in the string `"red, white, blue"`  then used str.split() to get the list of values.  Extreneous white space at the beginning or end needs stripping from pretty much any string read in from a csv.

**Comment by kcoyle** [comment]
I did something similar using excel. 

Input:

red white blue
red, white, blue
red,white,blue
"red""white""blue"
"red" "white"   "blue"
"red","white","blue"
red\|white\|blue

output:

red white blue
"red, white, blue"
"red,white,blue"
"""red""""white""""blue"""
"""red"" ""white"" ""blue"""
"""red"",""white"",""blue"""
red|white|blue

I read this as Excel quote-stringing any strings with commas, which then can be split, as per Phil. It doesn't add quotes where there are spaces or pipes. Mac Numbers gives the same results (if you ignore the extra columns). 

Should we include commas as possible delimiters when using a spreadsheet program?

**Comment by philbarker** [comment]
The current text says

> It is also advised that when creating a tabular profile in spreadsheet or table software, the contents of the cell, especially those containing commas, may be modified when exported to CSV format. You should consult the documentation for your software and test representative examples to ascertain how multiple cell values are exported.

I honestly think this is not the right emphasis. It makes the use of commas in spreadsheets sound more complicated than it is.  I think we have found that, while the CSV format requires special treatment of values containing commas or quotes[1], spreadsheets and other tools that handle import and export of CSVs generally deal with this without users having to think about it.  OTOH, any user creating or reading a tabular profile directly as a CSV will have to consult the CSV spec for how to deal with values that contain commas or quotes. 

1. https://tools.ietf.org/html/rfc4180

**Comment by kcoyle** [comment]
@philbarker  I had included that because our tests showed that the export to CSV was changed when commas were included but not with spaces or pipes. Re-reading this now, the export to CSV adds quotes but never really deals with multiple cell values. How about this:

It is also advised that when creating a tabular profile in spreadsheet or table software, the contents of the cell may be modified when exported to CSV format. You should consult the documentation for your software and test representative examples to ascertain if your multiple-value delimiter changes how cell values are exported.

**Comment by philbarker** [comment]
I still think the emphasis is wrong. The problem is that commas (and line feeds) need to be escaped in CSV, which is done with quotes, and so the quotes also need to be escaped. The easiest solution is to use a tool or library that handles this transparently. I would suggest:

It is advised that commas within values in a CSV file need to escaped so that they are not confused with commas that separate values. The CSV specification (https://tools.ietf.org/html/rfc4180) describes how to do this. However, most user-facing tools that are used edit CSV files, such as spreadsheets, handle this more or less transparently, as do many code libraries for handing CSV files programatically.

**Comment by kcoyle** [comment]
OK, I've edited that paragraph to read:

> Multiple values in a cell need to be delimited to distinguish them from a single value. Both the comma and the pipe character ("|") are commonly used delimiters that are highly visible within a string, but other characters may be used, with the caveat that the meaning of the characters used may need to be communicated to downstream users of the tabular profile. Note that comma characters are a special case in a CSV file, and commas used as multiple value delimiters need to escaped so that they are not confused with commas that separate columns. The CSV specification (https://tools.ietf.org/html/rfc4180) describes how to do this. However, most user-facing tools that are used edit CSV files, such as spreadsheets, handle this more or less transparently, as do many code libraries for processing CSV files programatically, so it often is not necessary to escape the commas when using a table or spreadsheet program. 

I used "commas that separate columns" because it seemed confusing to talk about commas separating multiple values in CSV vs commas separating values in a cell in a spreadsheet. (It's hard to be clear between spreadsheet function and CSV - I think this may need an explanatory section of its own in the primer. I fear that we haven't always made clear which we are describing. I'll try to read through the documents to check that.) 

Does this do it? Thanks.

**Comment by philbarker** [comment]
Yes, that's better. Thanks. +1 on clarity between what's needed in spreadsheets and what's needed in the raw file.

⁣Get TypeApp for Android ​

**Comment by tombaker** [comment]
The paragraph looks good, though it should perhaps be prefaced with some more general points: 
- the content of a cell (its "value") is by default interpreted as a single string
- in some columns, alternatively, a single cell may hold a list of multiple values
- there are different ways to format a cell with multiple values by using special punctuation and escape characters
- conventions differ according to whether one is entering values directly into a CSV file or using a spreadsheet program and exporting to CSV

The first sentence of your paragraph could perhaps be worked into this introductory paragraph.

Do we acknowledge somewhere that CSV-like formats can use delimiters other than commas, such as tabs or pipe characters?

One suggestion:
- `s/that handles this transparently/that handles this automatically`

**Comment by tombaker** [comment]
@kcoyle 
> It is also advised that when creating a tabular profile in spreadsheet or table software, the contents of the cell may be modified when exported to CSV format. 

I think it is good to make this point.

**Comment by kcoyle** [comment]
RE: CSV 

I think we need to make a few general statements about our intention:
1. We have created a vocabulary that is constrained such that it will work in a tabular model. However, it could also be used in a non-tabular scheme
2. When working in a tabular model, a common export format is "Comma Separated Values" (CSV). Other data formats are also used, such as tab-delimited values.

**Comment by philbarker** [comment]
> > It is also advised that when creating a tabular profile in spreadsheet or table software, the contents of the cell may be modified when exported to CSV format.

> I think it is good to make this point.

I think the wording creates unnecessary doubt. It sounds like the export might not be reliable, with entries changing in some unspecified way, when in fact it is handled quite smootlhly in a well defined way that most people don't need to worry about.

tbh, I think the hardest aspect of using commas to separate values in cells is finding language to talk about what happens  when they are exported as comma separate value files: there's a name collision when that has been noted above. Would it help if we spoke of items? E.g. "Multiple items in a cell need to be delimited..." (items seems appropriate for things in a list, perhaps less so for other places where we might have more than one entry in a cell). That seems better than finding an alternative to "values" when talking about CSVs, such as  "commas that separate columns", because commas separating values is kind of baked-in to CSV.

**Comment by kcoyle** [comment]
>I think the wording creates unnecessary doubt. 

@philbarker's comment is well taken. We can either leave this sentence off, or go for an explanation that speaks to the addition of quotation marks around some cell contents, in particular those containing commas. 

> "Multiple items in a cell need to be delimited

My problem with using "item" is that our columns have the word "value" in their names. We might be able to finesse this by referring to "a cell value that is a list of items...". I'd love to find something other than "item" though because that smacks of "things" not "strings". "terms"? "strings"? or just "a cell value that is to be processed as a list, not a single string" ?

**Comment by philbarker** [comment]
@kcoyle Yes, I share your concerns and reservations. Can we live with wording along the lines of "if the value is a list of options, then the options may separated with commas"?

**Comment by kcoyle** [comment]
Included in TAP primer Feb 13 2021
